src/lightning/pytorch/utilities/combined_loader.py

```diff
@@ -374,6 +375,24 @@ def _dataset_length(self) -> int:
         fn = _SUPPORTED_MODES[self._mode]["fn"]
         return fn(lengths)
 
+    def _state_dicts(self) -> List[Dict[str, Any]]:
+        """Returns the list of state dicts for iterables in `self.flattened` that are stateful."""
+        return [loader.state_dict() for loader in self.flattened if isinstance(loader, _Stateful)]
+
+    def _load_state_dicts(self, states: List[Dict[str, Any]]) -> None:
```

Comment:
Mainly because these are lists, and so the types don't match the `_Stateful` protocol. It seemed the right thing to do given that we use the `.flattened` representation (I opted for simplicity here). Would a different format make more sense?

Reply:
Good point. We could use the pytree spec to reconstruct the state dicts in the structure of the iterables, meaning returning a non-flattened view of the state dicts.

But we don't have a use for that, and nobody has requested this, so happy to keep it flattened and protected for now